### DISCUSSION TITLE
Minimal media for multiple samples vs single sample

### DISCUSSION THREAD

**Discussion by zoey-rw** [micom-dev/micom#177] [Q&A]
Hi, I was wondering how the minimal media is calculated when multiple community samples from the `build` workflow are passed as input to `fix_medium`. Will the resulting media file be able to reproduce the minimum observed growth rates on any of the samples? So it could have more metabolites than if any individual sample communities were passed to the workflow?

My "fixed" media files are smaller (75-100 metabolites) for a single sample and ~300 when I pass in a manifest with many samples. I'm wondering how to reconcile these media. Each community has up to 70 taxa and the other parameters are held constant (`community_growth=0.8, min_growth=0.01, max_import=1`, and the only metabolite in the candidate medium is glucose)

Thanks for any insight!

**Comment by zoey-rw** [comment]
It seems like this is exactly the situation that the `complete_db_medium` function was designed for, but `complete_db_medium` doesn't return the media pandas dataFrame that could be passed into the `grow` workflow. It does return the 300+ mets so those results are consistent!

**Comment by zoey-rw** [comment]
When using the second dataframe output from `complete_db_medium` and summing up the fluxes per metabolite as the medium file for `grow`, most communities couldn't grow (the flux values seemed too low)

I also tried getting the per-sample results using the `minimal_media` function with `summarize=False`, but I'm receiving this error after 100%:

```
from micom.workflows import minimal_media


minimal_media_persample = minimal_media(manual_amplicon_manifest, summarize=False,
                    model_folder="/projectnb/dietzelab/zrwerbin/N-cycle/data/MICOM/manual_amplicon/", 
                          min_growth=.01,threads=28)

Running ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:04:52

---------------------------------------------------------------------------
UnboundLocalError                         Traceback (most recent call last)
Cell In[5], line 3
      1 from micom.workflows import minimal_media
----> 3 minimal_media_persample = minimal_media(manual_amplicon_manifest, summarize=False,
      4                     model_folder="/projectnb/dietzelab/zrwerbin/N-cycle/data/MICOM/manual_amplicon/", 
      5                           min_growth=.01,threads=28)

File /projectnb/talbot-lab-data/zrwerbin/.conda/envs/micom/lib/python3.12/site-packages/micom/workflows/media.py:64, in minimal_media(manifest, model_folder, summarize, min_growth, threads)
     62 if summarize:
     63     medium = results.groupby("reaction").flux.max().reset_index()
---> 64 medium["metabolite"] = medium.reaction.str.replace("EX_", "")
     65 return medium

UnboundLocalError: cannot access local variable 'medium' where it is not associated with a value
```

**Comment by cdiener** [comment]
Hi,

You pretty much figured it out on your own already. `fix_medium` is mostly useful if you want to get nutrient requirements for individual samples, already having an idea of the overall or taxa growth rates but know nothing about the medium itself. `complete_db_medium` fixes a skeleton/template medium so that all organisms in a Model DB can grow on it. Growth rates in a community would then still be an emergent property (an actual prediction not an input like in `fix_medium`). 

For complete_db_medium there are a bunch of examples in the [media repository](https://github.com/micom-dev/media) (the recipes) that go through the individual steps. For instance [this one](https://nbviewer.org/github/micom-dev/media/blob/main/recipes/baobab_honey_antelope.ipynb) is a recent example that goes through a lot of curation steps. If you use the max of the import table from complete_db_medium the individual models should grow on that medium with the requested growth rate unless it failed (`can_grow` column in the manifest returned by complete_db_medium).
